File: modules/graphics/lib/rin/graphics/descriptors/DescriptorAllocator.cpp
#include "rin/graphics/descriptors/DescriptorAllocator.hpp"
#include "rin/core/GRuntime.hpp"
#include "rin/graphics/GraphicsModule.hpp"

void DescriptorAllocator::DestroyPools()
{
    for (auto& descriptorPool : _fullPools)
    {
        descriptorPool->Dispose();
    }

    for (auto& readyPool : _readyPools)
    {
        readyPool->Dispose();
    }

    _fullPools.clear();
    _readyPools.clear();

    _pools.clear();
}

void DescriptorAllocator::Reset()
{
    for (auto& readyPool : _readyPools)
    {
        readyPool->Reset();
    }

    for (auto& descriptorPool : _fullPools)
    {
        descriptorPool->Reset();
        _readyPools.emplace(descriptorPool);
    }

    _fullPools.clear();
}

Shared<DescriptorPool> DescriptorAllocator::GetPool()
{
    Shared<DescriptorPool> pool{};
    if (!_readyPools.empty())
    {
        pool = _pools[*_readyPools.begin()];
    }
    else
    {
        pool = CreatePool();
        _setsPerPool = static_cast<uint32_t>(_setsPerPool * 1.5);
        if (_setsPerPool > 4092) _setsPerPool = 4092;
        _readyPools.emplace(pool.get());
    }

    return pool;
}

Shared<DescriptorSet> DescriptorAllocator::Allocate(const vk::DescriptorSetLayout& layout,
                                                    const std::vector<uint32_t>& variableCount)
{
    auto targetPool = GetPool();
    Shared<DescriptorSet> set{};
    try
    {
        set = targetPool->Allocate(layout, variableCount);
    }
    catch (...)
    {
        _fullPools.emplace(targetPool.get());
        _readyPools.erase(targetPool.get());
        targetPool = GetPool();
        set = targetPool->Allocate(layout);
    }

    return set;
}

Shared<DescriptorPool> DescriptorAllocator::CreatePool()
{
    std::vector<vk::DescriptorPoolSize> poolSizes{};
    poolSizes.reserve(_ratios.size());
    for (auto& ratio : _ratios)
    {
        poolSizes.emplace_back(ratio.type, static_cast<uint32_t>(ratio.ratio * static_cast<float>(_setsPerPool)));
    }

    auto poolInfo = vk::DescriptorPoolCreateInfo{_poolCreateFlags, _setsPerPool, poolSizes};

    auto device = GRuntime::Get()->GetModule<GraphicsModule>()->GetDevice();
    auto pool = device.createDescriptorPool(poolInfo);
    auto sharedPool = newShared<DescriptorPool>(pool);
    _pools.emplace(sharedPool.get(), sharedPool);
    return sharedPool;
}

DescriptorAllocator::DescriptorAllocator(uint32_t maxSets, const std::vector<PoolSizeRatio>& poolRatios,
                                         const vk::DescriptorPoolCreateFlags& poolCreateFlags)
{
    _setsPerPool = maxSets;
    _ratios = poolRatios;
    _poolCreateFlags = poolCreateFlags;
}

Shared<DescriptorAllocator> DescriptorAllocator::New(uint32_t maxSets, const std::vector<PoolSizeRatio>& poolRatios,
                                                     const vk::DescriptorPoolCreateFlags& poolCreateFlags)
{
    return newShared<DescriptorAllocator>(maxSets, poolRatios, poolCreateFlags);
}

void DescriptorAllocator::OnDispose(bool manual)
{
    Disposable::OnDispose(manual);
    DestroyPools();
}

File: Rin.Engine/Graphics/Descriptors/DescriptorAllocator.cs
using TerraFX.Interop.Vulkan;
using static TerraFX.Interop.Vulkan.Vulkan;

namespace Rin.Engine.Graphics.Descriptors;

public class DescriptorAllocator : Disposable
{
    private readonly VkDevice _device;
    private readonly HashSet<DescriptorPool> _fullPools = [];
    private readonly VkDescriptorPoolCreateFlags _poolCreateFlags;
    private readonly PoolSizeRatio[] _ratios;
    private readonly HashSet<DescriptorPool> _readyPools = [];
    private uint _setsPerPool;

    public DescriptorAllocator(uint maxSets, PoolSizeRatio[] poolRatios,
        VkDescriptorPoolCreateFlags poolCreateFlags = 0)
    {
        _ratios = poolRatios;
        _setsPerPool = maxSets;
        _device = SGraphicsModule.Get().GetDevice();
        _poolCreateFlags = poolCreateFlags;
    }

    protected override void OnDispose(bool isManual)
    {
        DestroyPools();
    }

    public DescriptorSet Allocate(VkDescriptorSetLayout layout, params uint[] variableCount)
    {
        var targetPool = GetPool();
        DescriptorSet set;
        try
        {
            set = targetPool.Allocate(layout, variableCount);
        }
        catch (Exception)
        {
            _fullPools.Add(targetPool);
            _readyPools.Remove(targetPool);
            targetPool = GetPool();
            set = targetPool.Allocate(layout);
        }

        _readyPools.Add(targetPool);

        return set;
    }

    public void DestroyPools()
    {
        foreach (var pool in _fullPools) pool.Dispose();

        _fullPools.Clear();


        foreach (var pool in _readyPools) pool.Dispose();

        _readyPools.Clear();
    }

    public void ClearPools()
    {
        foreach (var pool in _readyPools) pool.Reset();

        foreach (var pool in _fullPools)
        {
            pool.Reset();
            _readyPools.Add(pool);
        }

        _fullPools.Clear();
    }

    private DescriptorPool GetPool()
    {
        DescriptorPool pool;
        if (_readyPools.Count > 0)
        {
            pool = _readyPools.First();
        }
        else
        {
            pool = CreatePool(_setsPerPool, _ratios);
            _setsPerPool = (uint)(_setsPerPool * 1.5);
            if (_setsPerPool > 4092) _setsPerPool = 4092;
            _readyPools.Add(pool);
        }

        return pool;
    }

    private unsafe DescriptorPool CreatePool(uint setCount, PoolSizeRatio[] poolRatios)
    {
        var poolSizes = poolRatios.Select(r =>
        {
            VkDescriptorPoolSize size = default;
            size.type = r.type.ToVk();
            size.descriptorCount = (uint)(r.ratio * setCount);
            return size;
        }).ToArray();


        var poolInfo = new VkDescriptorPoolCreateInfo
        {
            sType = VkStructureType.VK_STRUCTURE_TYPE_DESCRIPTOR_POOL_CREATE_INFO,
            flags = _poolCreateFlags,
            maxSets = setCount
        };
        fixed (VkDescriptorPoolSize* pPoolSizes = poolSizes)
        {
            poolInfo.pPoolSizes = pPoolSizes;
            poolInfo.poolSizeCount = (uint)poolSizes.Length;
        }

        VkDescriptorPool pool;
        vkCreateDescriptorPool(_device, &poolInfo, null, &pool);
        return new DescriptorPool(_device, pool);
    }
}